File: _attic/bane/core/marshal/optionals.go
package marshal

import (
	"reflect"

	"github.com/wrmsr/bane/core/check"
	rfl "github.com/wrmsr/bane/core/reflect"
	bt "github.com/wrmsr/bane/core/types"
)

///

type OptionalMarshaler struct {
	ty   reflect.Type
	elem Marshaler
}

func NewOptionalMarshaler(ty reflect.Type, elem Marshaler) OptionalMarshaler {
	return OptionalMarshaler{ty: ty, elem: elem}
}

var _ Marshaler = OptionalMarshaler{}

func (m OptionalMarshaler) Marshal(ctx *MarshalContext, rv reflect.Value) (Value, error) {
	iv := rv.Interface()
	o := iv.(bt.AnyOptional)
	if !o.Present() {
		return _nullValue, nil
	}

	return m.elem.Marshal(ctx, reflect.ValueOf(o.Interface()))
}

//

var _optionalInterfaceTy = rfl.TypeOf[bt.AnyOptional]()

var optionalMarshalerFactory = NewFuncFactory(func(ctx *MarshalContext, ty reflect.Type) (Marshaler, error) {
	if !ty.AssignableTo(_optionalInterfaceTy) {
		return nil, nil
	}

	ety := reflect.New(ty).Interface().(bt.AnyOptional).Type()
	check.NotNil(ety)
	elem, err := ctx.Make(ctx, ety)
	if err != nil {
		return nil, err
	}

	return NewOptionalMarshaler(ty, elem), nil
})

func NewOptionalMarshalerFactory() MarshalerFactory {
	return optionalMarshalerFactory
}

///

type OptionalUnmarshaler struct {
	ty   reflect.Type
	elem Unmarshaler

	nv  reflect.Value
	nvi bt.AnyOptional
}

func NewOptionalUnmarshaler(ty reflect.Type, elem Unmarshaler) OptionalUnmarshaler {
	nv := rfl.ZeroFor(ty)
	return OptionalUnmarshaler{ty: ty, elem: elem, nv: nv, nvi: nv.Interface().(bt.AnyOptional)}
}

var _ Unmarshaler = OptionalUnmarshaler{}

func (u OptionalUnmarshaler) Unmarshal(ctx *UnmarshalContext, mv Value) (reflect.Value, error) {
	switch mv := mv.(type) {

	case Null:
		return u.nv, nil

	default:
		ev, err := u.elem.Unmarshal(ctx, mv)
		if err != nil {
			return rfl.Invalid(), err
		}
		return reflect.ValueOf(u.nvi.Replace(ev.Interface())), nil
	}
}

//

var optionalUnmarshalerFactory = NewFuncFactory(func(ctx *UnmarshalContext, ty reflect.Type) (Unmarshaler, error) {
	if !ty.AssignableTo(_optionalInterfaceTy) {
		return nil, nil
	}

	ety := reflect.New(ty).Interface().(bt.AnyOptional).Type()
	check.NotNil(ety)
	elem, err := ctx.Make(ctx, ety)
	if err != nil {
		return nil, err
	}

	return NewOptionalUnmarshaler(ty, elem), nil
})

func NewOptionalUnmarshalerFactory() UnmarshalerFactory {
	return optionalUnmarshalerFactory
}

File: omlish/marshal/optionals.py
import dataclasses as dc
import typing as ta

from .. import reflect as rfl
from .base import MarshalContext
from .base import Marshaler
from .base import MarshalerFactory
from .base import UnmarshalContext
from .base import Unmarshaler
from .base import UnmarshalerFactory
from .values import Value


@dc.dataclass(frozen=True)
class OptionalMarshaler(Marshaler):
    e: Marshaler

    def marshal(self, ctx: MarshalContext, o: ta.Optional[ta.Any]) -> Value:
        if o is None:
            return None
        return self.e.marshal(ctx, o)


class OptionalMarshalerFactory(MarshalerFactory):
    def __call__(self, ctx: MarshalContext, rty: rfl.Type) -> ta.Optional[Marshaler]:
        if isinstance(rty, rfl.Union) and rty.is_optional:
            if (e := ctx.make(rty.without_none())) is None:
                return None  # type: ignore
            return OptionalMarshaler(e)
        return None


@dc.dataclass(frozen=True)
class OptionalUnmarshaler(Unmarshaler):
    e: Unmarshaler

    def unmarshal(self, ctx: UnmarshalContext, v: Value) -> ta.Optional[ta.Any]:
        if v is None:
            return None
        return self.e.unmarshal(ctx, v)


class OptionalUnmarshalerFactory(UnmarshalerFactory):
    def __call__(self, ctx: UnmarshalContext, rty: rfl.Type) -> ta.Optional[Unmarshaler]:
        if isinstance(rty, rfl.Union) and rty.is_optional:
            if (e := ctx.make(rty.without_none())) is None:
                return None  # type: ignore
            return OptionalUnmarshaler(e)
        return None
